app/snaputils.cpp

```diff
@@ -0,0 +1,203 @@
+  connect( mMapSettings, &QgsQuickMapSettings::outputSizeChanged, this, &SnapUtils::setup );
+  connect( mMapSettings, &QgsQuickMapSettings::outputDpiChanged, this, &SnapUtils::setup );
+
+  mSnappingUtils.setMapSettings( mMapSettings->mapSettings() );
+  emit mapSettingsChanged( mMapSettings );
+}
+
+void SnapUtils::getsnap( QPointF mapPoint )
+{
+  QgsPoint mapCoords = mMapSettings->screenToCoordinate( mapPoint );
+  QPointF snappoint = mCenterPosition; // by default show crosshair in center, no snap
+
+  // do no snap in the streaming mode
+  if ( !mUseSnapping )
+  {
+    setSnappedPosition( mCenterPosition );
+    setSnapped( false );
+    return;
+  }
+
+  QgsPointLocator::Match snap = mSnappingUtils.snapToMap( QgsPointXY( mapCoords.x(), mapCoords.y() ) );
+  if ( snap.isValid() )
+  {
+    snappoint = mMapSettings->coordinateToScreen( QgsPoint( snap.point().x(), snap.point().y() ) );
+    setSnappedPosition( snappoint );
+
+    if ( snap.hasVertex() )
+    {
+      setSnapType( SnapUtils::Vertex );
+    }
+    else if ( snap.hasArea() || snap.hasCentroid() || snap.hasMiddleSegment() )
+    {
+      setSnapType( SnapUtils::Other );
+    }
+    else
+    {
+      setSnapType( SnapUtils::Segment );
+    }
+
+    setSnapped( true );
+  }
+  else
+  {
+    setSnappedPosition( snappoint );
+    setSnapped( false );
+  }
+}
+
+void SnapUtils::setup()
+{
+  mSnappingUtils.setMapSettings( mMapSettings->mapSettings() );
+  setupSnapping();
+
+  getsnap( mCenterPosition );
+}
+
+QPointF SnapUtils::centerPosition() const
+{
+  return mCenterPosition;
+}
+
+void SnapUtils::setCenterPosition( QPointF newCenterPosition )
+{
+  if ( mCenterPosition == newCenterPosition )
+    return;
+  mCenterPosition = newCenterPosition;
+  emit centerPositionChanged( mCenterPosition );
+}
+
+QPointF SnapUtils::snappedPosition() const
+{
+  return mSnappedPosition;
+}
+
+void SnapUtils::setSnappedPosition( QPointF newSnappedPosition )
+{
+  if ( mSnappedPosition == newSnappedPosition )
+    return;
+  mSnappedPosition = newSnappedPosition;
+  emit snappedPositionChanged( mSnappedPosition );
+}
+
+bool SnapUtils::snapped() const
+{
+  return mSnapped;
+}
+
+void SnapUtils::setSnapped( bool newSnapped )
+{
+  if ( mSnapped == newSnapped )
+    return;
+  mSnapped = newSnapped;
+  emit snappedChanged( mSnapped );
+}
+
+const SnapUtils::SnapType &SnapUtils::snapType() const
+{
+  return mSnapType;
+}
+
+void SnapUtils::setSnapType( const SnapUtils::SnapType &newSnapType )
+{
+  if ( mSnapType == newSnapType )
+    return;
+  mSnapType = newSnapType;
+  emit snapTypeChanged( mSnapType );
+}
+
+bool SnapUtils::useSnapping() const
+{
+  return mUseSnapping;
+}
+
+void SnapUtils::setUseSnapping( bool useSnapping )
+{
+  mUseSnapping = useSnapping;
+}
+
+void SnapUtils::setupSnapping()
+{
+  int mode = mQgsProject->readNumEntry( QStringLiteral( "Mergin" ), QStringLiteral( "Snapping" ), 0 );
+  switch( mode )
+  {
+    case 0: break;
```

Comment:
Shouldn't this set `mUseSnapping` to false?

Reply:
Probably yes. And we should also handle this case (snapping in QGIS enabled, but disabled in Input) in the plugin and notify user. I will add another validation for it.